records/2021-09-27.batch-celery-save.md

```diff
@@ -0,0 +1,116 @@
+# Database performance issues
+
+Date: 2021-09-27
+
+## Status
+
+**PROPOSED.**
+
+## Context
+
+Following the ADR around scalability this previous July, one of the recommendation
+is to perform a batch save of the notifications when these are incoming into our
+API component rather than saving one by one. [A broader picture is portrayed in the
+ADR itself](https://github.com/cds-snc/notification-adr/blob/main/records/2021-07-16.scalability.database-bottleneck.md)
+with a larger context.
+
+An individual commit in a database in not a problem usually, but when
+several of these occurs, potentially over 100,000 in the case of Notify in time
+of high traffic, it becomes a performance issue.
+
+For a recap, imagine you are at the grocery store. You get to teh cashier to pay
+your items. You have to pay for each these items individually, i.e.
+a) scan one and only one item, b) ask for the total, c) pay with cash or card,
+d) repeat for all remaining items. This will take quite a long time compared
+to paying for all items in one batch within one transaction. Saving one
+notification every time in the database is akin to paying for one grocery item
+individually in one transaction, except you have potentially thousands of items
+remaining afterward. Ideally, we want to buy/save all for all of these in one
+go.
+
+### Current notification flow through Celery tasks
+
+> **INFO**: [Celery](https://docs.celeryproject.org/en/stable/getting-started/introduction.html) is a Python library to process a message through a workflow of tasks which offers reliability mechanism (e.g. retry), monitoring, concurrency (to execute multiple work at the same time), automatic scheduling of tasks and allows for ways to scale up the processing.
+
+Our current workflow of Celery tasks for processing an email or SMS at the moment
+is simplified with the next diagram.
+
+![Current workflow of a notification save](https://raw.githubusercontent.com/cds-snc/notification-adr/main/records/diagrams/2021-09-27.batch-celery-save/current-workflow.jpg?raw=true "Current workflow of a notification save")
```

Comment:
![current-workflow](https://user-images.githubusercontent.com/805567/134993152-3ec3cf29-7fac-4035-8611-5d2935e8c97c.jpg)

Reply:
What do the different colours/shapes represent? Is there a legend?

Reply:
nit: s/send-email-tasks queue/queue send-email-tasks/

Reply:
Is it a single retry queue or one per type of message? If it's a single one, would it make sense to draw as a single box?

Reply:
> Is it a single retry queue or one per type of message? If it's a single one, would it make sense to draw as a single box?

There is a single retry queue that backs up the three different and distinct Celery tasks. I would keep as-is because I didn't mean the boxes to represent the queue but rather the Celery tasks.